src/parser/mod.rs

```diff
@@ -10584,6 +10598,116 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn maybe_parse_table_sample(&mut self) -> Result<Option<Box<TableSampleMethod>>, ParserError> {
+        if self
+            .parse_one_of_keywords(&[Keyword::SAMPLE, Keyword::TABLESAMPLE])
+            .is_none()
+        {
+            return Ok(None);
+        }
+
+        // Try to parse based on an explicit table sample method keyword
+        let sample = if self
+            .parse_one_of_keywords(&[Keyword::BERNOULLI, Keyword::ROW])
+            .is_some()
+        {
+            self.expect_token(&Token::LParen)?;
+            let expr = self.parse_expr()?;
+
+            let (probability, value, unit) = if self.parse_keyword(Keyword::ROWS) {
+                (None, Some(expr), Some(TableSampleUnit::Rows))
+            } else if self.parse_keyword(Keyword::PERCENT) {
+                (None, Some(expr), Some(TableSampleUnit::Percent))
+            } else {
+                (Some(expr), None, None)
+            };
+            self.expect_token(&Token::RParen)?;
+            TableSampleMethod::Bernoulli(TableSampleBernoulli {
+                probability,
+                value,
+                unit,
+            })
+        } else if self
+            .parse_one_of_keywords(&[Keyword::SYSTEM, Keyword::BLOCK])
+            .is_some()
+        {
+            self.expect_token(&Token::LParen)?;
+            let probability = self.parse_expr()?;
+            self.expect_token(&Token::RParen)?;
+            let seed = if self
+                .parse_one_of_keywords(&[Keyword::REPEATABLE, Keyword::SEED])
+                .is_some()
+            {
+                self.expect_token(&Token::LParen)?;
+                let seed = self.parse_expr()?;
+                self.expect_token(&Token::RParen)?;
+                Some(seed)
+            } else {
+                None
+            };
+            TableSampleMethod::System(TableSampleSystem {
+                probability,
+                repeatable: seed,
+            })
+        // Try to parse without an explicit table sample method keyword
+        } else if self.peek_token().token == Token::LParen {
+            self.expect_token(&Token::LParen)?;
+            if self.parse_keyword(Keyword::BUCKET) {
+                let bucket = self.parse_number_value()?;
+                self.expect_keywords(&[Keyword::OUT, Keyword::OF])?;
+                let total = self.parse_number_value()?;
+                let on = if self.parse_keyword(Keyword::ON) {
+                    Some(self.parse_expr()?)
+                } else {
+                    None
+                };
+                self.expect_token(&Token::RParen)?;
+                TableSampleMethod::Bucket(TableSampleBucket { bucket, total, on })
+            } else {
+                let value = match self.maybe_parse(|p| p.parse_number_value()) {
+                    Ok(Some(num)) => num,
+                    _ => {
```

Comment:
```suggestion
                let value = match self.maybe_parse(|p| p.parse_number_value())? {
                    Some(num) => num,
                    None => {
```
I think this is usually the recursion limit or similar fatal error we can propagate